### PR TITLE
Robust GoDoc to document module operation and architecture

### PR DESCRIPTION
**Pull request by hannahhoward**
# Goals

This builds of @shannonwells efforts to document the Retrieval Module and shifts to documenting both the retrievalmarket and storagemarket modules through godoc

# Why GoDoc

Cause it's go's official way to document stuff!

But more importantly, documentation that doesn't live next to the code it documents gets out of date much quicker than docs you cant ignore. (especially if you're a responsible citizen and run 'golint' locally :)

# Implementation

- add missing public method documentation
- add package descriptions to all packages, including robust architecture documentation of the storage market and retrieval market modules
- breakup the storagemarket types file. It was so long!
- a couple fixes thrown in encountered while traversing lots of code
   - ProviderDealEnvironment's GetPieceSize method was ignoring the PieceCID even if specified
   - ListDeals methods were not implemented on RetrievalProvider or RetrievalClient
   - Make queryStream & dealStream implementations in RetrievalMarketNetwork private
   - Make RetrievalClient implementation a public struct (only one left that was private)

### CODE REVIEW

File: storagemarket/doc.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+https://filecoin-project.github.io/specs/#systems__filecoin_markets__storage_market
+
+The follow architectural components provide a brief overview of the design of
+the storagemarket module:
+
+Public Interfaces And Node Dependencies
+
+A core goal of this module is to isolate the negotation of deals from the actual chain operations
+performed by the node to put the deal on chain. The primarily orchestrates the storage deal
```
Comment (ingar) [inline storagemarket/doc.go:14]:
```suggestion
performed by the node to put the deal on chain. The module primarily orchestrates the storage deal
```

File: storagemarket/doc.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+flow, rather than performing specific chain operations which are delegated to the node.
+
+As such for both the client and the provider in the storage market, the module defines a top level
+public interface which it provides an implementation for, and a node interface that must be implemented
+by the filecoin node itself, and provided as a dependency. These node interfaces provide a universal way to
+talk to potentially multiple different filecoin node implementations, and can be implemented as using HTTP
```
Comment (ingar) [inline storagemarket/doc.go:20]:
```suggestion
talk to potentially multiple different filecoin node implementations, and can be implemented using HTTP
```

File: storagemarket/doc.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+or other interprocess communication to talk to a node implementation running in a different process
```
Comment (ingar) [inline storagemarket/doc.go:21]:
```suggestion
or other interprocess communication to talk to a node implementation running in a different process.
```

File: storagemarket/doc.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+node is expected to implement are StorageClientNode & StorageProviderNode. Further documentation of exactly what those
+dependencies should do can be found in the readme.
+
+Finite State Machines and Resumability
+
+Making deals in filecoin is a highly asynchronous process. For a large piece of data, it's possible from start
+to finish the proposing a deal, transferring data, publishing the deal, putting the data in a sector and sealing it
+could take hours or even days. Not surprisingly, many things can go wrong along the way. To manage the process
+of orchestrating deals, we use finite state machines that update deal state when discrete events occur. State updates
+always persist state to disk. This means we have a permaneant record of exactly what's going on with deals at any time,
```
Comment (ingar) [inline storagemarket/doc.go:33]:
```suggestion
always persist state to disk. This means we have a permanent record of exactly what's going on with deals at any time,
```

**Comment by ingar** [comment]
GFC had a nice (but kinda annoying) linter that would check for documentation on exported symbols, make sure that errors were formatted properly, etc.  Does this PR add that check?

**Comment by hannahhoward** [comment]
@ingar the linter you are thinking of is `golint` -- we could enable it in `golintci` (the meta linter that runs on circle) -- it has no configuration options and no ability to skip lines... which is a bummer... and the docs say "Don't run this on CI"...

My VSCode runs it locally so I get lots of messages. We could enable it as a non-required step.

**Pull request merged**